### PR TITLE
IMPROVE - Adding a ProgressView to DashboardView

### LINKED ISSUE #108: IMPROVE - Adding a `ProgressView` to `DashboardView`
## Describe the feature you want to improve
* When the data is being fetched from Firebase, show a `ProgressView` once it's loaded the data will be shown in a list like normal
* Make sure the empty view overlay is still shown after returning 0 results

There would be 3 states:
* Empty, no results
* List, showing one or more `MaintenanceEvent`s
* `ProgressView`, loading data from Firebase

### PR DESCRIPTION
**Pull request by mariaykim**
# What it Does
* Closes #108 
* Creating a `LoadingState` enum to support different views per loading state for the Dashboard View

# How I Tested
* Build the app
* Verify the spinner shows upon app launch
* Navigate away from the Dashboard and reclick the Dashboard tab
* Verify the spinner shows before the desired result displays

# Notes

# Screenshot

| No Results | Has Results | Error Handling |
| ---- | --- | --- |
| https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/80354714/eb1f3dde-95df-4fd3-a81e-96e3329ff0f4 | https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/80354714/99fdc269-c302-49c2-9de4-0bb18267e55b | https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/80354714/c797b978-484f-4f3f-ba18-7b0dd118b27e |

### CODE REVIEW

**Comment by mikaelacaron** [comment]
@mariaykim Please solve the merge conflicts, thank you!

File: Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift [hunk 1]
```diff
@@ -106,14 +75,16 @@ struct DashboardView: View {
...
             .sheet(isPresented: $isShowingAddView) {
                 makeAddMaintenanceView()
             }
         }
+        .onAppear {
+            Task {
+                await viewModel.getMaintenanceEvents()
+                await viewModel.getVehicles()
+            }
+        }
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift]: Why is this changed from being in `.task` to being in `.onAppear`?

File: Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift [hunk 2]
```diff
@@ -179,4 +183,9 @@ extension DashboardViewModel {
             }
         }
     }
+    
+    enum LoadingState: String {
+        case isLoading
+        case loaded
+    }
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift]: Move this to the top of the file, not at the bottom

File: Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift [hunk 3]
```diff
@@ -66,6 +68,7 @@ class DashboardViewModel {
     }
     
     func getMaintenanceEvents() async {
+        loadingState = .isLoading
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift:78]: I haven't tested this yet, but what happens when an error is thrown, is this forever "loading?"

**Review by mikaelacaron — CHANGES_REQUESTED** [review]
Thanks for contributing! I have a couple of questions!

File: Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift (hunk 1 above)
Comment (mariaykim) [reply Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift]: I had difficulties with `.task` during development, but I can change it back.

File: Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift (hunk 3 above)
Comment (mariaykim) [reply Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift:78]: Yes, the loading spinner would continue to show.

Comment (mikaelacaron) [reply Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift:78]: Does the alert show?

File: Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift (hunk 1 above)
Comment (mikaelacaron) [reply Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift]: What issues?

Comment (mariaykim) [reply Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift]: I didn't have real issues, so we should be able to keep `.task`

File: Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift (hunk 3 above)
Comment (mariaykim) [reply Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift:78]: I added an error state to the `LoadingState` enum and created a new error alert to show.

**Comment by mikaelacaron** [comment]
I'll be looking at this tonight! Can you update the branch and fix merge conflicts please @mariaykim

File: Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift [hunk 4]
```diff
@@ -9,6 +9,12 @@ import FirebaseFirestore
 import FirebaseFirestoreSwift
 import Foundation
 
+enum LoadingState: String {
+    case isLoading
+    case loaded
+    case error
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift:15]: I don't think an `error` state is needed, this would indicate that the data has finished loading. What I was thinking was an `isLoading` boolean value

File: Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift [hunk 5]
```diff
@@ -17,6 +23,7 @@ class DashboardViewModel {
     var events = [MaintenanceEvent]()
     var showAddErrorAlert = false
     var showErrorAlert = false
+    var showFailedLoadErrorAlert = false
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift:26]: Another error later variable is not needed. Use the existing `showErrorAlert`

File: Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift [hunk 6]
```diff
@@ -82,6 +92,11 @@ class DashboardViewModel {
                     }
                 }
                 self.events = events
+                self.loadingState = .loaded
+            } else {
+                showFailedLoadErrorAlert.toggle()
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift:97]: note because of what I mentioned earlier, use the existing variables as opposed to the new ones

File: Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift [hunk 7]
```diff
@@ -135,6 +106,51 @@ struct DashboardView: View {
         }
     }
     
+    private var maintenanceListView: some View {
```
Comment (mikaelacaron) [inline Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift:109]: Also I think for right now, it's fine to leave this where it was, and not make a variable of it, just yet

**Review by mikaelacaron — CHANGES_REQUESTED** [review]
Thanks for your patience! While this works, and you noted in your PR comment that you need to go to the `DashboardView`, tap a different tab and then come back, I think that's actually a separate bug of something happening. (I'll make a new issue that anyone can look at for this).

I have several comments that need resolved please. I was thinking you only need an `isLoading` boolean property to accomplish this issue. I don't think a 3rd enum state is beneficial.

If you have any questions, you can comment directly on that thread, otherwise click the resolve button once you've addressed my comments. Once you fix everything, please click the re-review button

Re-review button

**Comment by mikaelacaron** [comment]
@mariaykim  Thank you for contributing!

I'm going to be closing this PR and un-assigning you from this issue because it hasn't been updated in 3 days, and there's only 5 days left in Hacktoberfest, and I want to allow others to be able to contribute to the project. You're more than welcome to contribute to this project on a different issue, or after Hacktoberfest
